### ISSUE TITLE
Can we avoid RecordStream for P2P ops?

### ISSUE THREAD

**Issue by wconstab** [pytorch/pytorch#132292]
**Background: RecordStream is a safety/convenience feature that sacrifices some performance**

RecordStream is a feature that automates safety of cuda-allocator reuse/free when tensors are used across multiple nccl streams.  TLDR; If the tensor is used on a separate stream from which it's allocated, the allocator could attempt to free it too early, and recordStream implements a cuda-event-based callback system to ensure the remote stream work compeltes before allocator reuse/free can occur.  Why is this bad?  It can lead to non-determinism of memory reuse, that can lead to significant performance variations in distributed programs.  Often times, we'd rather be explicit about freeing, so that we know when it will happen and it happens at the same time on all ranks.

**Disabling RecordStream**
Today there exists a flag (`TORCH_NCCL_AVOID_RECORD_STREAMS`) for disabling recordStream in much of the ProcessGroupNCCL code, but P2P ops are excepted from this for the following reason: [ProcessGroupNCCL::pointToPoint()](https://github.com/pytorch/pytorch/blob/main/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L2910)

> "for isend, I don't think the API requires the user to wait() on the returned handle, so ProcessGroupNCCL can't know when it's safe to release the input"

To be clear, the isend API returns a work object, created inside the pointToPoint function, which holds a reference to the input tensor.  Holding this reference should keep the tensor alive until the work object is out of scope.

**Questions**

Calling work.wait() causes the current stream to wait for the communication (e.g. send) to finish before it lets additional ops run, which is important if for example additional ops were to write into the send buffer.   But is the caching allocator actually aware of this wait event, or is it free to free the send buffer as soon as the work object goes out of scope (and thus the tensor loses its refcount)?

**Comment by fduwjj** [comment]
If users don't hold the tensor long enough for send to finish, would this also lead to a collective hang?

**Comment by awgu** [comment]
I do not think it would hang, but it could have incorrect results. Another kernel could write to the send input while the send is still happening.
